File: ios/Godstone/Sources/GodstoneMesh/Frame.swift
import Foundation

/// GMP/1 wire frame. Byte-for-byte identical to the Android implementation in
/// tab 04. Big-endian throughout. See 02_MESH_PROTOCOL section 5.
///
///   0      1        2        3        4 .. 19      20 .. 23   24 .. 25   26 ..
///   ver    type     ttl      flags    msg_id(16)   routing    len(2)     payload
public struct Frame: Sendable, Equatable {

    public static let version: UInt8 = 1
    public static let headerSize = 26
    public static let maxPayload = 60_000

    public enum FrameType: UInt8, Sendable {
        case handshake = 0x01
        case sos       = 0x02
        case direct    = 0x03
        case group     = 0x04
        case broadcast = 0x05
        case bulk      = 0x06
        case digest    = 0x07
        case ack       = 0x08

        /// Delivery order under congestion. SOS always wins.
        var priority: Int {
            switch self {
            case .handshake: return 0
            case .sos:       return 1
            case .ack:       return 2
            case .direct:    return 3
            case .digest:    return 4
            case .group:     return 5
            case .broadcast: return 6
            case .bulk:      return 7
            }
        }
    }

    public struct Flags: OptionSet, Sendable {
        public let rawValue: UInt8
        public init(rawValue: UInt8) { self.rawValue = rawValue }

        public static let sealed      = Flags(rawValue: 1 << 0)
        public static let compressed  = Flags(rawValue: 1 << 1)
        public static let fragmented  = Flags(rawValue: 1 << 2)
        public static let hasProofOfWork = Flags(rawValue: 1 << 3)
        public static let requiresAck = Flags(rawValue: 1 << 4)
    }

    public let type: FrameType
    public let ttl: UInt8
    public let flags: Flags
    public let messageId: Data      // 16 bytes
    public let routingTag: Data     // 4 bytes, rotates daily
    public let payload: Data

    public init(type: FrameType, ttl: UInt8, flags: Flags,
                messageId: Data, routingTag: Data, payload: Data) {
        precondition(messageId.count == 16, "message id must be 16 bytes")
        precondition(routingTag.count == 4, "routing tag must be 4 bytes")
        precondition(payload.count <= Frame.maxPayload, "payload too large")
        self.type = type
        self.ttl = ttl
        self.flags = flags
        self.messageId = messageId
        self.routingTag = routingTag
        self.payload = payload
    }

    public func encode() -> Data {
        var out = Data(capacity: Frame.headerSize + payload.count)
        out.append(Frame.version)
        out.append(type.rawValue)
        out.append(ttl)
        out.append(flags.rawValue)
        out.append(messageId)
        out.append(routingTag)
        out.append(UInt8((payload.count >> 8) & 0xFF))
        out.append(UInt8(payload.count & 0xFF))
        out.append(payload)
        return out
    }

    /// Bounded parsing. Every length is validated against what actually remains
    /// in the buffer before a single byte is allocated. A malicious peer must
    /// not be able to make us allocate 60 KB by lying in a 2-byte field.
    public static func decode(_ data: Data) throws -> Frame {
        guard data.count >= headerSize else { throw MeshError.malformedFrame }
        let b = [UInt8](data)

        guard b[0] == version else { throw MeshError.malformedFrame }
        guard let type = FrameType(rawValue: b[1]) else { throw MeshError.malformedFrame }

        let declared = (Int(b[24]) << 8) | Int(b[25])
        guard declared <= maxPayload,
              data.count == headerSize + declared else {
            throw MeshError.malformedFrame
        }

        return Frame(
            type: type,
            ttl: b[2],
            flags: Flags(rawValue: b[3]),
            messageId: data.subdata(in: 4..<20),
            routingTag: data.subdata(in: 20..<24),
            payload: data.subdata(in: headerSize..<(headerSize + declared))
        )
    }

    public func decremented() -> Frame? {
        guard ttl > 1 else { return nil }
        return Frame(type: type, ttl: ttl - 1, flags: flags,
                     messageId: messageId, routingTag: routingTag, payload: payload)
    }
}

File: android/mesh/src/main/java/io/godstone/mesh/wire/Frame.kt
package io.godstone.mesh.wire

import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * GMP/1 frame. See docs/mesh/PROTOCOL.md section 5.
 *
 * Layout (big-endian):
 *   0   1   version
 *   1   1   type
 *   2   2   length
 *   4   1   ttl
 *   5   1   priority
 *   6   8   msg_id
 *   14  6   timestamp
 *   20  N   payload
 */
data class Frame(
    val version: Byte = PROTOCOL_VERSION,
    val type: FrameType,
    val ttl: Int,
    val priority: Priority,
    val msgId: Long,
    val timestamp: Long,
    val payload: ByteArray
) {

    fun encode(): ByteArray {
        require(payload.size <= MAX_PAYLOAD) { "payload too large: ${payload.size}" }
        val buf = ByteBuffer.allocate(HEADER_SIZE + payload.size)
            .order(ByteOrder.BIG_ENDIAN)
        buf.put(version)
        buf.put(type.code)
        buf.putShort(payload.size.toShort())
        buf.put(ttl.coerceIn(0, MAX_TTL).toByte())
        buf.put(priority.code)
        buf.putLong(msgId)
        // 6-byte timestamp: build 8 bytes then keep the low 6.
        val ts = ByteBuffer.allocate(8).order(ByteOrder.BIG_ENDIAN)
            .putLong(timestamp).array()
        buf.put(ts, 2, 6)
        buf.put(payload)
        return buf.array()
    }

    companion object {
        const val PROTOCOL_VERSION: Byte = 0x01
        const val HEADER_SIZE = 20
        const val MAX_PAYLOAD = 65535
        const val MAX_TTL = 16
        const val DEFAULT_TTL = 12

        /**
         * Bounded parsing, protocol section 8. Every length field is validated
         * against the actual remaining buffer BEFORE any allocation, so an
         * attacker-supplied length can never drive memory allocation.
         */
        fun decode(raw: ByteArray): Frame? {
            if (raw.size < HEADER_SIZE) return null

            val buf = ByteBuffer.wrap(raw).order(ByteOrder.BIG_ENDIAN)

            val version = buf.get()
            if (version != PROTOCOL_VERSION) return null   // refuse, never guess

            val type = FrameType.from(buf.get()) ?: return null
            val length = buf.getShort().toInt() and 0xFFFF

            // Critical bound check.
            if (length != raw.size - HEADER_SIZE) return null
            if (length > MAX_PAYLOAD) return null

            val ttl = buf.get().toInt() and 0xFF
            if (ttl > MAX_TTL) return null

            val priority = Priority.from(buf.get()) ?: return null
            val msgId = buf.getLong()

            val tsBytes = ByteArray(8)
            buf.get(tsBytes, 2, 6)
            val timestamp = ByteBuffer.wrap(tsBytes).order(ByteOrder.BIG_ENDIAN).getLong()

            val payload = ByteArray(length)
            buf.get(payload)

            return Frame(version, type, ttl, priority, msgId, timestamp, payload)
        }
    }

    override fun equals(other: Any?): Boolean =
        other is Frame && other.msgId == msgId

    override fun hashCode(): Int = msgId.hashCode()
}

enum class FrameType(val code: Byte) {
    HELLO(0x01),
    DIGEST(0x02),
    WANT(0x03),
    MESSAGE(0x04),
    ACK(0x05),
    BULK_OFFER(0x06),
    BULK_CHUNK(0x07),
    SOS(0x08),
    PING(0x09),
    GOODBYE(0x0A);

    companion object {
        private val map = entries.associateBy { it.code }
        fun from(b: Byte): FrameType? = map[b]
    }
}

enum class Priority(val code: Byte) {
    SOS(0),
    DIRECT(1),
    GROUP(2),
    BROADCAST(3),
    BULK(4);

    /** SOS and DIRECT are exempt from proof of work: latency there is safety. */
    val requiresProofOfWork: Boolean get() = this == GROUP || this == BROADCAST

    companion object {
        private val map = entries.associateBy { it.code }
        fun from(b: Byte): Priority? = map[b]
    }
}
